infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesPersistentVolumeClaims.java

```diff
@@ -231,10 +218,56 @@ public void onClose(KubernetesClientException cause) {
       }
     } catch (KubernetesClientException e) {
       throw new KubernetesInfrastructureException(e);
-    } finally {
-      if (watch != null) {
-        watch.close();
-      }
     }
   }
+
+  private Watch pvcIsBoundWatcher(
+      CompletableFuture<PersistentVolumeClaim> future,
+      Resource<PersistentVolumeClaim, DoneablePersistentVolumeClaim> pvcResource,
+      Predicate<PersistentVolumeClaim> predicate,
+      String name) {
+    return pvcResource.watch(
+        new Watcher<PersistentVolumeClaim>() {
+          @Override
+          public void eventReceived(Action action, PersistentVolumeClaim pvc) {
+            if (predicate.test(pvc)) {
+              future.complete(pvc);
+            }
+          }
+
+          @Override
+          public void onClose(KubernetesClientException cause) {
+            future.completeExceptionally(
+                new InfrastructureException(
+                    "Waiting for persistent volume claim '" + name + "' was interrupted"));
+          }
+        });
+  }
+
+  private Watch pvcIsWaitingForConsumerWatcher(
+      CompletableFuture<PersistentVolumeClaim> future, PersistentVolumeClaim actualPvc)
+      throws InfrastructureException {
+    return clientFactory
+        .create(workspaceId)
+        .events()
```

Comment:
Please note that old events may be received as well.
Maybe it makes sense to filter events initializing... But it's needed to be careful here since an event may be published after PVC creating but before subscribing.
See https://github.com/eclipse/che/blob/d1b9203099b553866ee5a13b2645c71fdb6f07ff/infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesDeployments.java#L496

So, not sure that we should do it here, maybe it's not an issue... Probably we should listen to events only after first workspace run, during all other - PVC should be already bound and we should exit later.
Just be aware of such behavior

Reply:
Playing with timestamps is very tricky, not reliable and usually indication that you're doing something wrong. I don't like the solution you're referring. If you know how to reproduce this situation, I can test and find solution for that.

Reply:
>  If you know how to reproduce this situation, I can test and find solution for that.

well. as I mentioned 

> So, not sure that we should do it here, maybe it's not an issue... Probably we should listen to events only after the first workspace run, during all other - PVC should be already bound and we should exit later.

But I can invent some case for testing:
1. Configure `common` or `per-workspace` PVC strategy.
2. Configure storage class with `waitForFirstConsumer`.
3. Start Che, create and start any workspace, then stop it.
4. Remove workspace related PVC.
5. Configure storage class without `waitForFirstConsumer`.
6. Restart Che and start the same workspace again. Probably Che Server won't wait until PVC is bound because it will receive `waitForFirstConsumer` event.

It's a crazy scenario. You can close it and just resolve the comment :smile_cat: